python/TestHarness/resultsstore/reader.py

```diff
@@ -201,85 +324,85 @@ def iterateLatestPushResults(
                 if isinstance(num, int):
                     batch_size = 1
 
-                for data in cursor:
-                    id = data.get("_id")
-                    assert isinstance(id, ObjectId)
-
-                    # So that we know where to search next time
-                    self._last_latest_push_event_id = id
+                built_results = [self._build_result(doc) for doc in docs]
+                self._latest_push_results += built_results
+                if built_results:
+                    self._last_latest_push_event_id = built_results[-1].id
 
-                    # Due to invalidation, we could have multiple results
-                    # from the same event. If we already have this event
-                    # (the latest version of it), skip this one
-                    event_id = data.get("event_id")
-                    if event_id is not None and any(
-                        r.event_id == event_id for r in self._latest_push_results
-                    ):
-                        continue
+            # Have a shared result to yield
+            if i < len(self._latest_push_results):
+                result = self._latest_push_results[i]
 
-                    # Build/get the result and store it for future use
-                    result = self._buildResults(data)
-                    self._latest_push_results.append(result)
+                # Make sure IDs are decreasing
+                if i > 1:
+                    assert result.id < self._latest_push_results[i - 1].id
 
-            if i < len(self._latest_push_results):
-                yield self._latest_push_results[i]
+                yield result
                 i += 1
+            # No more results, and we didn't find any more
             else:
+                self._found_final_push_event = True
```

Comment:
Is this supposed to be reset somewhere? Seems like the loop would break early if the the iterator finished once and was called again.

Reply:
It's supposed to exit early if that's the case. If we find the end (the oldest result), there's never anything else to find.